index.bs

```diff
@@ -1142,10 +1149,15 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
     [=biometric recognition=] (e.g., presenting a fingerprint) [[ISOBiometricVocabulary]]. The intent is to
     distinguish individual users.
 
+    Note that [=user verification=] does not give the [=[RP]=] a concrete identification of the user,
+    only an assertion that multiple instances of [=user verification=] for the same [=credential=] are performed by the same user.
```

Comment:
Multiple?

Reply:
How about "all instances of (successful?) [=user verification=] have been performed by the same user"?

> but when you have done 2 or more UV ceremonies with that credential it expresses that it was the same user that performed all of them.

Just to clarify a subtle point: Would this guarantee about the user also be expected to apply in the case where the credential was created without a UV ceremony, but is then used multiple times to generate an assertion with UV requested?

Reply:
>but when you have done 2 or more UV ceremonies with that credential it expresses that it was the same user that performed all of them.

I like this idea, thanks!

>Would this guarantee about the user also be expected to apply in the case where the credential was created without a UV ceremony, but is then used multiple times to generate an assertion with UV requested?

Yes, that is my intent.

Reply:
> Yes, that is my intent.

I asked for clarification on this point because I believe that it could unnecessarily limit future authenticator designs. For example, on a high-powered authenticator (like a phone with a TEE or SE), UV would probably be tracked by binding an individual credential created with UV requested to the current user at the time or even their concrete biometric information. Introducing a meaningful user binding after creation time is often not possible or at least not well supported in hardware by existing key store implementations (like the Android Keystore or iOS Keychain). If UV were not a flag, but instead a cryptographic fingerprint like the one provided by the "uvi" extension, this would be less of an issue.

It is also not clear to me what kind of application could make use of the guarantee that the same user requested multiple assertions for a credential that may have been created by someone else. Is this something you could see a potential use case for? If not, it may be worthwhile to make the guarantee here slightly more restrictive in order to not complicate authenticators in an unnecessary way.

Reply:
> It is also not clear to me what kind of application could make use of the guarantee that the same user requested multiple assertions for a credential that may have been created by someone else.

Indeed, that's not likely to be an explicit use case. I wrote this mostly to reflect how most roaming authenticators work: you can create credentials without setting up UV (often by PIN) on the authenticator, but once you have UV set up you can't disable it without destroying all your credential keys.

So it's not so much a use case as an emergent consequence of how implementations interact. It's also not unthinkable that an RP might start out with `userVerification: "discouraged"`, but later decide to move more towards passwordless MFA and change to `userVerification: "preferred"`. In that case they probably have a bunch of users whose credentials were created without UV. It doesn't really matter if UV wasn't set up when the credential was created - the RP can still use UV in place of a password for future ceremonies, as long as the _first_ UV ceremony is also authenticated by the password (or some other second factor).

> Introducing a meaningful user binding after creation time is often not possible or at least not well supported in hardware by existing key store implementations (like the Android Keystore or iOS Keychain).

Ok, but then that sounds to me like they wouldn't be able to support UV for credentials created without UV, correct? I.e., they'd be skipped in [step 18 of `get()`](https://www.w3.org/TR/2019/WD-webauthn-2-20190604/#ref-for-dom-publickeycredentialrequestoptions-userverification):

>If `options.userVerification` is set to `required` and the authenticator is not capable of performing user verification, continue.

Reply:
> So it's not so much a use case as an emergent consequence of how implementations interact. It's also not unthinkable that an RP might start out with `userVerification: "discouraged"`, but later decide to move more towards passwordless MFA and change to `userVerification: "preferred"`. In that case they probably have a bunch of users whose credentials were created without UV. It doesn't really matter if UV wasn't set up when the credential was created - the RP can still use UV in place of a password for future ceremonies, as long as the _first_ UV ceremony is also authenticated by the password (or some other second factor).

This is a situation I hadn't thought about, thanks for the example. I agree that this is a situation in which allowing to ascertain UV at a later point in time is advantageous.

> Ok, but then that sounds to me like they wouldn't be able to support UV for credentials created without UV, correct? I.e., they'd be skipped in step 18 of get():

So far I have taken this step to not depend on the registered credentials, but only care about whether the authenticator is configured to support UV (e.g. has a fingerprint registered) at the moment. I may have misunderstood this part of the spec though, please let me know if it's supposed to say something else.

My expectation would rather have been that the authenticator would simply not return the UV flag for credentials registered without UV, regardless of whether UV was required or not. Would this be a good way for authenticators to behave if they cannot track users of non-UV keys, but still support UV in general?

Reply:
> So far I have taken this step to not depend on the registered credentials, but only care about whether the authenticator is configured to support UV (e.g. has a fingerprint registered) at the moment. I may have misunderstood this part of the spec though, please let me know if it's supposed to say something else.

I don't quite understand what you mean here. WebAuthn ceremonies are all about credentials, you can't do anything without one.

> My expectation would rather have been that the authenticator would simply not set the UV flag in GetAssertion responses for credentials created without UV, even if UV is required for the corresponding request. Would this be a good way for authenticators to behave if they cannot track users of non-UV keys, but still support UV in general?

Yes, that's exactly right. The UV flag in the authenticator response indicates whether UV was used in this particular ceremony. If UV=1, then the authenticator must of course have UV configured, but UV=0 does not imply that the authenticator is not configured or not capable of UV.

Reply:
> Yes, that's exactly right. The UV flag in the authenticator response indicates whether UV was used in this particular ceremony. If UV=1, then the authenticator must of course have UV configured, but UV=0 does not imply that the authenticator is not configured or not capable of UV.

Thanks for the confirmation, this clears up the potential issues I saw with the user-tracking guarantee explained in this PR as authenticators have a way to gracefully decline tracking a user if the credential was created without UV. Additionally, being allowed to return UV=0 even if the request has set UV to required helps with #1306 to some extent.

> I don't quite understand what you mean here. WebAuthn ceremonies are all about credentials, you can't do anything without one.

I apologize for not being clear enough in my reply. I just wanted to remark that the part of the spec you referenced is part of the steps to take "If an authenticator becomes available on this client device,", and hence the information retrieved from the authenticator in this step would probably only depend on the general state and configuration of the authenticator rather than the properties of a particular credential. But this doesn't really matter anymore thanks to your clarification above.

Reply:
> Additionally, being allowed to return UV=0 even if the request has set UV to required helps with #1306 to some extent.

Actually, this isn't allowed - if the request sets `userVerification: "required"` then the authenticator must perform UV or fail. It is fine to return UV=0 for `userVerification: "preferred"`, though.

Reply:
> Actually, this isn't allowed - if the request sets `userVerification: "required"` then the authenticator must perform UV or fail. It is fine to return UV=0 for `userVerification: "preferred"`, though.

Since the mapping of `userVerification: preferred` to either `true` or `false` will loose this information before it reaches the authenticator, this may unfortunately not help much. Thank you very much for all the clarifications though!

It seems to me that a "true" solution to this and other issues (#1306 + the somewhat hacky U2F fallback in Chrome, and also the client not being able to judge whether UV is "low-effort" on the authenticator) would be to pass the full trivalent `userVerification` set by the RP through to the authenticator via an extension. I would gladly work on this if you should find it a reasonable proposal, but I can understand if this is not a priority because current authenticator designs would not be able to make use of it.

Reply:
Ok. Let's close this discussion, then. You're welcome to open a new issue for further clarifications/improvements.